File: backend/game/RhelGamePython/chapter2_create_config_fs/autofs_client.py
import slow_validInput


def configure_autofs_server():
    """
    Install and configure autofs for NFS server on Red Hat.
    """
    try:
        slow_validInput.print_slow(
            "\n\nConfiguring autofs for NFS Server:\n\n")
        slow_validInput.print_slow(
            "As you survey the landscape of Auto-q652-FS, you realize that optimizing Autofs configurations on the server side will be crucial.")
        slow_validInput.print_slow(
            "You access the central server, the heart of the planet's filesystem management system, and begin to configure Autofs settings.\n")
        slow_validInput.print_slow(
            "You define automount maps, specifying the filesystems to be automatically mounted when accessed by users or applications.\n")
        slow_validInput.print_slow(
            "Each map is meticulously crafted to ensure efficient data access and management across the planet's network.\n")
        slow_validInput.print_slow(
            "Additionally, you fine-tune timeout settings, ensuring that idle filesystems are unmounted promptly to conserve resources.\n")
        slow_validInput.print_slow(
            "With each configuration tweak, the server becomes more responsive and efficient, seamlessly managing filesystems to meet the demands of Auto-q652-FS.\n")

        slow_validInput.print_slow(
            "You can quit at any time by typing 'quit' or 'q'.\n")

        # Define the correct commands in the specified order
        correct_commands = [
            # Appends NFS export configuration to /etc/exports
            "echo '/home/nfsServer ClientIP/24(rw,no_root_squash,no_subtree_check)' >> /etc/exports",
            # Restarts NFS server to apply changes in /etc/exports
            "systemctl restart nfs-server.service",
            "exportfs -a",  # Exports all directories listed in /etc/exports
            "showmount -e localhost"  # Shows NFS shares exported by the localhost
        ]
        to_do = [
            "Append NFS export configuration to /etc/exports",
            "Restarts NFS server to apply changes in /etc/exports",
            "Exports all directories listed in /etc/exports",
            "Shows NFS shares exported by the localhost",
        ]

        quit_commands = ["quit", "q"]
        index = 0
        slow_validInput.print_slow(
            "There are 4 commands you must give in the correct order!")
        while index < len(correct_commands):
            user_command = input(
                f"{to_do[index]}. Enter the command: '{index + 1}' (type 'quit' or 'q' to exit): ")
            if user_command.strip() in quit_commands:
                slow_validInput.print_slow("Exiting the task. Farewell!")
                return False

            if user_command.strip() == correct_commands[index]:
                slow_validInput.print_slow("Executing the command...")
                # Here you can execute the command using subprocess or other methods
                slow_validInput.print_slow(
                    f"Command '{correct_commands[index]}' executed successfully!\n")
                # slow_validInput.print_slow explanations for each command
                if index == 0:
                    slow_validInput.print_slow(
                        "The 'echo' command appends the specified NFS export configuration to /etc/exports.")
                    slow_validInput.print_slow(
                        "This configuration allows the NFS server to share the directory '/home/nfsServer' with the IP address 'ClientIP'.")
                elif index == 1:
                    slow_validInput.print_slow(
                        "The 'systemctl restart nfs-server.service' command restarts the NFS server service.")
                    slow_validInput.print_slow(
                        "This is done to apply the changes made to the NFS exports file (/etc/exports).")
                elif index == 2:
                    slow_validInput.print_slow(
                        "The 'exportfs -a' command exports all directories listed in /etc/exports.")
                    slow_validInput.print_slow(
                        "This makes the NFS shares available for mounting by remote clients.")
                elif index == 3:
                    slow_validInput.print_slow(
                        "The 'showmount -e localhost' command displays the NFS shares exported by the localhost.")
                    slow_validInput.print_slow(
                        "This confirms that the NFS server is successfully exporting the specified directories.")

                index += 1
            else:
                slow_validInput.print_slow("Incorrect command. Try again.")
                slow_validInput.print_slow(
                    f"Hint: Use '{correct_commands[index]}' to proceed.")
                continue

        slow_validInput.print_slow(
            "Autofs installation and configuration completed successfully!\n")
        return True

    except KeyboardInterrupt:
        slow_validInput.print_slow(
            "\nExiting the task due to user interruption (Ctrl+C). Farewell!")
        return False
    except Exception as e:
        slow_validInput.print_slow("An error occurred:", e)
        return False


def configure_autofs_client():
    """
    Configure autofs on the client side.
    """
    try:
        slow_validInput.print_slow(
            "\n\nConfiguring autofs for NFS Client:\n\n")
        slow_validInput.print_slow(
            "With the Autofs configurations optimized on the server side, it's time to configure Autofs on the client side.")
        slow_validInput.print_slow(
            "You access the client machines scattered across Auto-q652-FS, ready to implement the necessary changes.\n")
        slow_validInput.print_slow(
            "On each client machine, you edit the Autofs configuration files, defining the mount points and specifying the server addresses.\n")
        slow_validInput.print_slow(
            "Each client is carefully configured to seamlessly access the filesystems managed by the central server.\n")
        slow_validInput.print_slow(
            "You verify the configurations, ensuring that Autofs is set to automatically mount the required filesystems upon access.\n")

        slow_validInput.print_slow(
            "To proceed, follow the steps below to configure autofs on your system:\n")
        slow_validInput.print_slow(
            "You can quit at any time by typing 'quit' or 'q'.\n")

        correct_commands = [
            "dnf install -y autofs",  # Installs autofs package
            "mkdir autofs",  # Creates a directory to mount the NFS shares
            # Adds autofs configuration to /etc/auto.master
            "echo 'autofs    /etc/auto.sharedfs    --timeout=30' >> /etc/auto.master",
            # Copies the default autofs configuration file
            "cp /etc/auto.misc /etc/auto.sharedfs",
            # Adds NFS share configuration to /etc/auto.sharedfs
            "echo 'nfsServer -fstype=nfs    ServerIP:/home/user1/nfsServer' >> /etc/auto.sharedfs",
            "systemctl restart autofs.service",  # Restarts autofs service to apply changes
            "ls /autofs"  # Checks if the directory to be created is mounted
        ]
        to_do = [
            "Install autofs package",
            "Create a directory to mount the NFS shares",
            "Add autofs configuration to /etc/auto.master",
            "Copy the default autofs configuration file",
            "Add NFS share configuration to /etc/auto.sharedfs",
            "Restarts autofs service to apply changes",
            "Checks if the directory to be created is mounted",
        ]
        quit_commands = ["quit", "q"]
        index = 0
        slow_validInput.print_slow(
            "There are 7 commands you must give in the correct order!")
        while index < len(correct_commands):
            user_command = input(
                f"{to_do[index]}. Enter the command: '{index + 1}' (type 'quit' or 'q' to exit): ")
            if user_command.strip() in quit_commands:
                slow_validInput.print_slow("Exiting the task. Farewell!")
                return False

            if user_command.strip() == correct_commands[index]:
                slow_validInput.print_slow("Executing the command...")
                slow_validInput.print_slow(
                    f"Command '{correct_commands[index]}' executed successfully!\n")
                # slow_validInput.print_slow explanations for each command
                explanations = [
                    "The 'dnf install -y autofs' command installs the autofs package.",
                    "The 'mkdir autofs' command creates a directory to mount the NFS shares.",
                    "The 'echo' command adds autofs configuration to /etc/auto.master.",
                    "The 'cp' command copies the default autofs configuration file to /etc/auto.sharedfs.",
                    "The 'echo' command adds NFS share configuration to /etc/auto.sharedfs.",
                    "The 'systemctl restart autofs' command restarts the autofs service to apply changes.",
                    "The 'ls /autofs' command checks if the directory to be created is mounted.",
                ]
                slow_validInput.print_slow(
                    f"Explanation: {explanations[index]}\n")
                index += 1
            else:
                slow_validInput.print_slow("Incorrect command. Try again.")
                slow_validInput.print_slow(
                    f"Hint: Use '{correct_commands[index]}' to proceed.")
                continue

        slow_validInput.print_slow(
            "Autofs configuration on the client side completed successfully!\n")
        return True

    except KeyboardInterrupt:
        slow_validInput.print_slow(
            "\nExiting the task due to user interruption (Ctrl+C). Farewell!")
        return False
    except Exception as e:
        slow_validInput.print_slow("An error occurred:", e)
        return False


def install_nfs():
    """
    Install NFS (Network File System) on Red Hat.
    """
    try:
        slow_validInput.print_slow(
            "\n\nInstalling NFS (Network File System):\n\n")
        slow_validInput.print_slow(
            "With Autofs optimizations complete, you turn your attention to the next phase of infrastructure development: NFS (Network File System) installation and configuration.")
        slow_validInput.print_slow(
            "NFS is essential for enabling seamless file sharing and access across the network, a crucial component for the interconnected systems of Auto-q652-FS.\n")
        slow_validInput.print_slow(
            "You access the central server, the nucleus of Auto-q652-FS's data management network, to begin the NFS installation process.\n")
        slow_validInput.print_slow(
            "Using advanced cybernetic tools, you install the NFS packages and dependencies, laying the foundation for networked file sharing.\n")
        slow_validInput.print_slow(
            "Once the installation is complete, you proceed to configure the NFS exports, defining the directories and permissions to be shared across the network.\n")
        slow_validInput.print_slow(
            "Each export is meticulously crafted to ensure secure and efficient access to the shared filesystems, maintaining data integrity and confidentiality.\n")

        slow_validInput.print_slow(
            "You can quit at any time by typing 'quit' or 'q'.\n")

        correct_command = "dnf install -y nfs-utils"
        quit_commands = ["quit", "q"]

        while True:
            user_command = input(
                "Enter the command to install NFS (type 'quit' or 'q' to exit): ")
            if user_command.strip() in quit_commands:
                slow_validInput.print_slow("Exiting the task. Farewell!")
                return False

            if user_command.strip() == correct_command:
                slow_validInput.print_slow(
                    "The command resonates with power as you utter it aloud, invoking the installation process.")
                slow_validInput.print_slow(
                    "Mystical energies swirl around you as NFS utilities are installed,")
                slow_validInput.print_slow(
                    "their capabilities now at your command.\n")
                slow_validInput.print_slow("NFS installed successfully!\n")
                slow_validInput.print_slow("Output Example:")
                slow_validInput.print_slow("  Installed:\n"
                                           "    nfs-utils.x86_64\n")
                return True
            else:
                slow_validInput.print_slow("Incorrect command. Try again.")
                slow_validInput.print_slow(
                    "Hint: Use 'dnf install -y nfs-utils' to install NFS.")
                continue
    except KeyboardInterrupt:
        slow_validInput.print_slow(
            "\nExiting the task due to user interruption (Ctrl+C). Farewell!")
        return False
    except Exception as e:
        slow_validInput.print_slow("An error occurred:", e)
        return False

File: backend/data/chapters/chapter2/challenges/missions/autofs_client.js
// backend/data/chapters/chapter2/missions/autofs_client.js

const autofs_client = [
  {
    id: "configure_autofs_server",
    title: "Configure Autofs on Server Side",
    action: "configure NFS server with autofs for dynamic file mounting",
    intro: `
As you survey the landscape of Auto-q652-FS, you realize that optimizing Autofs configurations on the server side will be crucial.
You access the central server, the heart of the planet's filesystem management system, and begin to configure Autofs settings.

You define automount maps, specifying the filesystems to be automatically mounted when accessed by users or applications.

Each map is meticulously crafted to ensure efficient data access and management across the planet's network.

Additionally, you fine-tune timeout settings, ensuring that idle filesystems are unmounted promptly to conserve resources.

With each configuration tweak, the server becomes more responsive and efficient, seamlessly managing filesystems to meet the demands of Auto-q652-FS.
    `,
    solution: "multi_command_sequence",
    hints: [
      "This requires a sequence of 4 commands in the correct order.",
      "Commands involve: NFS export configuration, service restart, export activation, and verification."
    ],
    output: `NFS Server Configuration Sequence:
1. NFS export configuration appended to /etc/exports
2. NFS server service restarted to apply changes
3. All directories in /etc/exports exported successfully
4. NFS shares verified and displayed`,
    aspects: [
      "- echo '/home/nfsServer ClientIP/24(rw,no_root_squash,no_subtree_check)' >> /etc/exports: Appends NFS export configuration",
      "- systemctl restart nfs-server.service: Restarts NFS server to apply changes",
      "- exportfs -a: Exports all directories listed in /etc/exports",
      "- showmount -e localhost: Shows NFS shares exported by the localhost"
    ],
    options: [
      "- NFS export permissions: rw (read-write), no_root_squash (preserve root privileges), no_subtree_check (disable subtree checking)",
      "- Service management: systemctl commands for managing NFS services",
      "- Export verification: showmount command to verify active NFS exports"
    ],
    outro: `As you complete the Autofs configurations on the server side, you take pride in knowing that you have optimized the planet's data infrastructure to its fullest potential.

With your mission on Auto-q652-FS accomplished, you prepare to embark on new adventures in the boundless expanse of the cosmos.

The NFS server now stands ready to serve filesystems across the network with unprecedented efficiency and reliability.`,
  },

  {
    id: "configure_autofs_client",
    title: "Configure Autofs on Client Side",
    action: "configure autofs client for seamless NFS access",
    intro: `
With the Autofs configurations optimized on the server side, it's time to configure Autofs on the client side.
You access the client machines scattered across Auto-q652-FS, ready to implement the necessary changes.

On each client machine, you edit the Autofs configuration files, defining the mount points and specifying the server addresses.

Each client is carefully configured to seamlessly access the filesystems managed by the central server.

You verify the configurations, ensuring that Autofs is set to automatically mount the required filesystems upon access.
    `,
    solution: "multi_command_sequence",
    hints: [
      "This requires a sequence of 7 commands in the correct order.",
      "Commands involve: package installation, directory creation, configuration files, and service management."
    ],
    output: `Autofs Client Configuration Sequence:
1. Autofs package installed successfully
2. Mount directory created for NFS shares
3. Autofs master configuration updated
4. Default configuration file copied
5. NFS share configuration added
6. Autofs service restarted
7. Mount directory verified and accessible`,
    aspects: [
      "- dnf install -y autofs: Installs autofs package",
      "- mkdir autofs: Creates a directory to mount the NFS shares",
      "- echo 'autofs /etc/auto.sharedfs --timeout=30' >> /etc/auto.master: Adds autofs configuration",
      "- cp /etc/auto.misc /etc/auto.sharedfs: Copies the default autofs configuration file",
      "- echo 'nfsServer -fstype=nfs ServerIP:/home/user1/nfsServer' >> /etc/auto.sharedfs: Adds NFS share configuration",
      "- systemctl restart autofs.service: Restarts autofs service to apply changes",
      "- ls /autofs: Checks if the directory to be created is mounted"
    ],
    options: [
      "- Timeout configuration: --timeout=30 sets unmount timeout for idle filesystems",
      "- Filesystem type: -fstype=nfs specifies NFS as the filesystem type",
      "- Service management: systemctl commands for managing autofs service"
    ],
    outro: `As you complete the Autofs configurations on the client side, you witness a seamless integration of filesystems across Auto-q652-FS.

Users and applications can now access data with unprecedented ease, thanks to your expertise in configuring Autofs.

With the Autofs configurations perfected on both the server and client sides, Auto-q652-FS stands ready to embrace the future of data management.

As you prepare to depart for new adventures, you reflect on the impact of your actions, knowing that you have ushered Auto-q652-FS into a new era of efficiency and innovation.`,
  },

  {
    id: "install_nfs",
    title: "Install NFS (Network File System)",
    action: "install NFS utilities for network file sharing",
    intro: `
With Autofs optimizations complete, you turn your attention to the next phase of infrastructure development: NFS (Network File System) installation and configuration.
NFS is essential for enabling seamless file sharing and access across the network, a crucial component for the interconnected systems of Auto-q652-FS.

You access the central server, the nucleus of Auto-q652-FS's data management network, to begin the NFS installation process.

Using advanced cybernetic tools, you install the NFS packages and dependencies, laying the foundation for networked file sharing.

Once the installation is complete, you proceed to configure the NFS exports, defining the directories and permissions to be shared across the network.

Each export is meticulously crafted to ensure secure and efficient access to the shared filesystems, maintaining data integrity and confidentiality.
    `,
    solution: "dnf install -y nfs-utils",
    hints: [
      "Hint: Use 'dnf install -y nfs-utils' to install NFS utilities.",
      "The answer is: dnf install -y nfs-utils"
    ],
    output: `Installed:
  nfs-utils.x86_64
Complete!`,
    aspects: [
      "- dnf install -y: Install packages with automatic yes confirmation",
      "- nfs-utils: Network File System utilities package"
    ],
    options: [
      "- NFS utilities: Provides tools for NFS server and client operations",
      "- Network file sharing: Enables sharing of filesystems across network",
      "- Cross-platform compatibility: Works with various operating systems"
    ],
    outro: `The command resonates with power as you utter it aloud, invoking the installation process.
Mystical energies swirl around you as NFS utilities are installed, their capabilities now at your command.

As the NFS exports are configured, you test the connectivity from client machines, verifying seamless access to the shared directories.

With NFS successfully installed and configured, Auto-q652-FS is now equipped with a robust network file sharing infrastructure, enabling seamless collaboration and data access across its interconnected systems.

As you prepare to depart for new adventures, you take pride in knowing that your expertise has strengthened the foundations of Auto-q652-FS, ensuring its continued prosperity in the vast expanse of the cosmos.`,
  }
];

export default autofs_client;